# Setup & Imports

In [1]:
# ------------------ Data Manipulation ------------------
import pandas as pd          # DataFrame operations
import numpy as np           # Numerical computations

# ------------------ Visualization ------------------
import matplotlib.pyplot as plt  # Basic plotting
from mpl_toolkits.mplot3d import Axes3D  # 3D plotting support
import seaborn as sns             # Statistical visualization

# ------------------ Statistics & Analysis ------------------
from scipy.stats import pearsonr    # Pearson correlation
import scipy.stats as stats         # General statistical tools
from statsmodels.stats.outliers_influence import variance_inflation_factor  # VIF

# ------------------ Machine Learning & Scaling ------------------
from sklearn.preprocessing import MinMaxScaler, StandardScaler  # Normalization and standardization
from sklearn.cluster import KMeans                              # KMeans clustering
from sklearn.preprocessing import PolynomialFeatures            # Polynomial basis for regression
from sklearn.linear_model import LinearRegression               # Linear regression model
import pandas as pd
import numpy as np
import xgboost as xgb
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split, RandomizedSearchCV
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import OneHotEncoder
import joblib
import warnings
import json
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.linear_model import Ridge
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR
from xgboost import XGBRegressor
from sklearn.inspection import PartialDependenceDisplay
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score
# ------------------ Optimization & Pareto Analysis ------------------
from pymoo.util.nds.non_dominated_sorting import NonDominatedSorting  # Pareto front identification

# ------------------ Geospatial Analysis ------------------
import geopandas as gpd         # Handling shapefiles and spatial data
import fiona                    # I/O support for geospatial formats
import pyogrio                  # Fast I/O for vector geodata
import rasterio                 # Raster data reading/writing
from rasterstats import zonal_stats  # Zonal statistics from rasters

# ------------------ Utility & System ------------------
import os                # File system interaction
import warnings          # Suppress warnings
warnings.filterwarnings("ignore")  # Ignore all warnings

# ------------------ Pandas Display Settings ------------------
pd.options.display.float_format = '{:.2f}'.format         # Format float output
pd.set_option('display.max_columns', None)                # Show all columns

# ------------------ Matplotlib Settings ------------------
plt.rcParams['font.sans-serif'] = ['Arial']              # Support Chinese characters
plt.rcParams['axes.unicode_minus'] = False                # Display minus signs correctly

# Inline plot display (only needed in notebooks)
%matplotlib inline

from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import OneHotEncoder
import xgboost as xgb
import joblib
import json

import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor, ExtraTreesRegressor
from sklearn.svm import SVR
from sklearn.neighbors import KNeighborsRegressor
from xgboost import XGBRegressor
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import train_test_split

from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR
from xgboost import XGBRegressor

from sklearn.metrics import r2_score, mean_squared_error
from sklearn.model_selection import train_test_split
import numpy as np
import pandas as pd
# from autogluon.tabular import TabularDataset,TabularPredictor

# Data import & preprocess

In [2]:
df = pd.read_csv('../data/grid500_36cities_with_eff_k500.csv')
grid500 = df.copy()
print(grid500.shape)
grid500.head()

(194941, 52)


Global_ID     City  NTL2023  NTL2023_focal  VIT202311  UHIDAY2020_07_inv  \
0          0  Beijing     6.18           7.03    1649.86               0.00   
1          1  Beijing     5.74           7.23     155.86               0.00   
2          2  Beijing    11.32          13.62      57.14               0.00   
3          3  Beijing    11.73          14.36     243.57               0.00   
4          4  Beijing     8.45           9.19       0.00               0.00   

   buildingCount  totalHeight  avgHeight  maxHeight  minHeight  heightRange  \
0             31        93.00       3.00       3.00       3.00         0.00   
1             29        87.00       3.00       3.00       3.00         0.00   
2             22        66.00       3.00       3.00       3.00         0.00   
3              8        24.00       3.00       3.00       3.00         0.00   
4              0         0.00       0.00       0.00       0.00         0.00   

   heightIndex  heightDensity  heightVariety  cornerCountTotal  \
0         0.03           0.00           0.00               180   
1         0.03           0.00           0.00               149   
2         0.05           0.00           0.00               117   
3         0.12           0.00           0.00                42   
4         0.00           0.00           0.00                 0   

   basePerimeterTotal  basePerimeterAvg  basePerimeterMax  basePerimeterMin  \
0             6500.98            209.71            831.63             32.55   
1             3810.39            131.39            272.14             15.33   
2             3645.40            165.70            668.26             28.73   
3             1147.60            143.45            283.01             75.66   
4                0.00              0.00              0.00              0.00   

   shapeComplexity  compactness  baseAreaTotal  totalArea  areaVariance  \
0             5.07         0.56       88704.58   88704.58   50961857.70   
1             5.17         0.62       33119.49   33119.49    1261729.88   
2             4.92         0.56       51092.86   51092.86   30358745.83   
3             5.70         0.60        8534.61    8534.61     958399.89   
4             0.00         0.00           0.00       0.00          0.00   

   buildingIntensity  avgBuildingArea  parcelArea  largestPatchIndex  \
0               0.35          2861.44   132998.25               0.47   
1               0.13          1142.05   196625.22               0.67   
2               0.20          2322.40   132197.31               0.53   
3               0.03          1066.83   205685.37               0.82   
4               0.00             0.00       99.15               0.00   

   shape3DIndex  SVF  balanceIndex  FAR  coverageRatio  FVC  permeableRatio  \
0          0.01 0.80        234.60 0.35           0.35 0.84            0.89   
1          0.26 0.89         35.66 0.13           0.13 0.86            0.94   
2          0.01 0.90        151.50 0.20           0.20 0.88            0.93   
3          0.02 0.95         15.54 0.03           0.03 0.93            0.98   
4          0.00 0.98          0.00 0.00           0.00 0.82            0.98   

   poiDensity  poiDiversity  streetRatio  roadDensity  intersectionDensity  \
0           4          0.00         0.20            5                    1   
1          12          0.64         0.20            7                    1   
2           0          0.00         0.20            1                    0   
3           0          0.00         0.20            3                    1   
4          16          0.69         0.00            1                    1   

   cornerCountAvg  buildingProximity  buildingMinDist  buildingMaxDist  \
0            5.81              52.56            14.99           620.94   
1            5.14              34.22             6.82           594.16   
2            5.32              47.90            17.34           610.84   
3            5.25              35.93            13.49    

In [3]:
grid500.describe()

Global_ID   NTL2023  NTL2023_focal  VIT202311  UHIDAY2020_07_inv  \
count  194941.00 189742.00      194941.00  194941.00          194941.00   
mean    98137.51     28.88          28.65    1414.14              -0.99   
std     56862.94     19.36          17.87    1821.23               1.00   
min         0.00      0.00           0.00       0.00              -4.60   
25%     49033.00     15.06          15.64     206.86              -1.69   
50%     97800.00     25.85          26.01     731.14              -0.91   
75%    147135.00     38.67          38.17    2003.43               0.00   
max    196769.00    465.96         348.37   68494.71               5.13   

       buildingCount  totalHeight  avgHeight  maxHeight  minHeight  \
count      194941.00    194941.00  194941.00  194941.00  194941.00   
mean           46.70       893.91      16.53      30.42       9.58   
std            50.96      1035.48       8.72      24.26       4.91   
min             0.00         0.00       0.00       0.00       0.00   
25%            12.00       144.00      12.00      12.00       6.00   
50%            35.00       559.00      16.67      27.00      12.00   
75%            65.00      1315.00      21.99      40.00      12.00   
max          1189.00     22354.50     130.50     535.00     112.00   

       heightRange  heightIndex  heightDensity  heightVariety  \
count    194941.00    194941.00      194941.00      194941.00   
mean         20.83         0.10           0.00           0.30   
std          23.99         0.17           0.00           0.28   
min           0.00         0.00           0.00           0.00   
25%           0.00         0.02           0.00           0.00   
50%          15.00         0.04           0.00           0.28   
75%          28.00         0.09           0.01           0.43   
max         532.00         1.00           0.09           2.87   

       cornerCountTotal  basePerimeterTotal  basePerimeterAvg  \
count         194941.00           194941.00         194941.00   
mean             315.42             4959.32            112.06   
std              352.35             4245.40             60.43   
min                0.00                0.00              0.00   
25%               72.00             1417.38             83.02   
50%              222.00             4200.02            110.75   
75%              445.00             7512.92            137.30   
max             8619.00            50912.47           1689.25   

       basePerimeterMax  basePerimeterMin  shapeComplexity  compactness  \
count         194941.00         194941.00        194941.00    194941.00   
mean             347.82             22.65             4.55         0.54   
std              230.34             35.73             1.40         0.17   
min                0.00              0.00             0.00         0.00   
25%              202.45              4.35             4.57         0.53   
50%              319.04             14.07             4.82         0.57   
75%              461.66             28.38             5.08         0.62   
max             5146.21           1689.25            39.49         0.99   

       baseAreaTotal  totalArea    areaVariance  buildingIntensity  \
count      194941.00  194941.00       194941.00          194941.00   
mean        33192.68  216494.27    191423545.70               0.87   
std         27474.13  210010.02   2300870309.71               0.84   
min             0.00       0.00            0.00               0.00   
25%          8888.17   37241.51      3519241.57               0.15   
50%         29057.71  157756.72     20726892.83               0.63   
75%         51784.69  347326.83     74777609.98               1.39   
max        226640.62 3672930.95 662000000000.00              14.69   

       avgBuildingArea  parcelArea  largestPatchIndex  shape3DIndex       SVF  \
count        194941.00   194941.00          194941.00     194941.00 194941.00   
mean           5437.01   180006.93               0.50      

## Urban form indicators

In [4]:
features = [
    'buildingCount', 'totalHeight', 'avgHeight', 'basePerimeterTotal',
    'basePerimeterAvg', 'compactness', 'baseAreaTotal', 'totalArea',
    'avgBuildingArea', 'FAR', 'coverageRatio',
    'roadDensity', 'intersectionDensity', 'poiDensity', 'poiDiversity',
    'FVC', 'permeableRatio', 'maxHeight', 'minHeight', 'heightRange', 'heightIndex',
    'heightDensity', 'heightVariety', 'cornerCountTotal', 'basePerimeterMax',
    'basePerimeterMin', 'shapeComplexity', 'areaVariance', 'parcelArea',
    'largestPatchIndex', 'shape3DIndex', 'balanceIndex', 'cornerCountAvg',
    'buildingProximity', 'buildingMinDist', 'buildingMaxDist',
    'buildingDistAvg', 'buildingDistVar', 'SVF', 'streetRatio'
]

In [5]:
# Extract the selected feature subset
df_feat = grid500[features].copy()

# Identify skewed features for transformation
skewness = df_feat.skew()
skewed_cols = skewness[skewness > 1].index.tolist()  # Skewness > 1 = strongly right-skewed
print(skewed_cols)

['buildingCount', 'totalHeight', 'basePerimeterTotal', 'basePerimeterAvg', 'totalArea', 'avgBuildingArea', 'FAR', 'roadDensity', 'intersectionDensity', 'poiDensity', 'maxHeight', 'heightRange', 'heightIndex', 'heightDensity', 'heightVariety', 'cornerCountTotal', 'basePerimeterMax', 'basePerimeterMin', 'areaVariance', 'shape3DIndex', 'balanceIndex', 'cornerCountAvg', 'buildingProximity', 'buildingMinDist', 'buildingDistVar', 'streetRatio']


In [6]:
len(skewed_cols)

26

In [7]:
# Create copy for transformation
grid500_transformed = grid500.copy()

# Define a small epsilon shift to ensure positivity
epsilon = 1e-6  # or you can use slightly larger like 1e-3

for col in skewed_cols:
    min_val = grid500_transformed[col].min()
    if min_val <= 0:
        shift = abs(min_val) + epsilon
        print(f"ℹ️ Shifting {col} by +{shift:.2e} to ensure positivity before log1p.")
        grid500_transformed[f'log_{col}'] = np.log1p(grid500_transformed[col] + shift)
    else:
        grid500_transformed[f'log_{col}'] = np.log1p(grid500_transformed[col])

grid500_transformed.describe()

ℹ️ Shifting buildingCount by +1.00e-06 to ensure positivity before log1p.
ℹ️ Shifting totalHeight by +1.00e-06 to ensure positivity before log1p.
ℹ️ Shifting basePerimeterTotal by +1.00e-06 to ensure positivity before log1p.
ℹ️ Shifting basePerimeterAvg by +1.00e-06 to ensure positivity before log1p.
ℹ️ Shifting totalArea by +1.00e-06 to ensure positivity before log1p.
ℹ️ Shifting avgBuildingArea by +1.00e-06 to ensure positivity before log1p.
ℹ️ Shifting FAR by +1.00e-06 to ensure positivity before log1p.
ℹ️ Shifting roadDensity by +1.00e-06 to ensure positivity before log1p.
ℹ️ Shifting intersectionDensity by +1.00e-06 to ensure positivity before log1p.
ℹ️ Shifting poiDensity by +1.00e-06 to ensure positivity before log1p.
ℹ️ Shifting maxHeight by +1.00e-06 to ensure positivity before log1p.
ℹ️ Shifting heightRange by +1.00e-06 to ensure positivity before log1p.
ℹ️ Shifting heightIndex by +1.00e-06 to ensure positivity before log1p.
ℹ️ Shifting heightDensity by +1.00e-06 to ensure po

Global_ID   NTL2023  NTL2023_focal  VIT202311  UHIDAY2020_07_inv  \
count  194941.00 189742.00      194941.00  194941.00          194941.00   
mean    98137.51     28.88          28.65    1414.14              -0.99   
std     56862.94     19.36          17.87    1821.23               1.00   
min         0.00      0.00           0.00       0.00              -4.60   
25%     49033.00     15.06          15.64     206.86              -1.69   
50%     97800.00     25.85          26.01     731.14              -0.91   
75%    147135.00     38.67          38.17    2003.43               0.00   
max    196769.00    465.96         348.37   68494.71               5.13   

       buildingCount  totalHeight  avgHeight  maxHeight  minHeight  \
count      194941.00    194941.00  194941.00  194941.00  194941.00   
mean           46.70       893.91      16.53      30.42       9.58   
std            50.96      1035.48       8.72      24.26       4.91   
min             0.00         0.00       0.00       0.00       0.00   
25%            12.00       144.00      12.00      12.00       6.00   
50%            35.00       559.00      16.67      27.00      12.00   
75%            65.00      1315.00      21.99      40.00      12.00   
max          1189.00     22354.50     130.50     535.00     112.00   

       heightRange  heightIndex  heightDensity  heightVariety  \
count    194941.00    194941.00      194941.00      194941.00   
mean         20.83         0.10           0.00           0.30   
std          23.99         0.17           0.00           0.28   
min           0.00         0.00           0.00           0.00   
25%           0.00         0.02           0.00           0.00   
50%          15.00         0.04           0.00           0.28   
75%          28.00         0.09           0.01           0.43   
max         532.00         1.00           0.09           2.87   

       cornerCountTotal  basePerimeterTotal  basePerimeterAvg  \
count         194941.00           194941.00         194941.00   
mean             315.42             4959.32            112.06   
std              352.35             4245.40             60.43   
min                0.00                0.00              0.00   
25%               72.00             1417.38             83.02   
50%              222.00             4200.02            110.75   
75%              445.00             7512.92            137.30   
max             8619.00            50912.47           1689.25   

       basePerimeterMax  basePerimeterMin  shapeComplexity  compactness  \
count         194941.00         194941.00        194941.00    194941.00   
mean             347.82             22.65             4.55         0.54   
std              230.34             35.73             1.40         0.17   
min                0.00              0.00             0.00         0.00   
25%              202.45              4.35             4.57         0.53   
50%              319.04             14.07             4.82         0.57   
75%              461.66             28.38             5.08         0.62   
max             5146.21           1689.25            39.49         0.99   

       baseAreaTotal  totalArea    areaVariance  buildingIntensity  \
count      194941.00  194941.00       194941.00          194941.00   
mean        33192.68  216494.27    191423545.70               0.87   
std         27474.13  210010.02   2300870309.71               0.84   
min             0.00       0.00            0.00               0.00   
25%          8888.17   37241.51      3519241.57               0.15   
50%         29057.71  157756.72     20726892.83               0.63   
75%         51784.69  347326.83     74777609.98               1.39   
max        226640.62 3672930.95 662000000000.00              14.69   

       avgBuildingArea  parcelArea  largestPatchIndex  shape3DIndex       SVF  \
count        194941.00   194941.00          194941.00     194941.00 194941.00   
mean           5437.01   180006.93               0.50      

In [8]:
log_features = [f'log_{col}' for col in skewed_cols]
non_log_features = [col for col in features if col not in skewed_cols]

features_transformed = log_features + non_log_features

print(len(features_transformed))
features_transformed

40


['log_buildingCount',
 'log_totalHeight',
 'log_basePerimeterTotal',
 'log_basePerimeterAvg',
 'log_totalArea',
 'log_avgBuildingArea',
 'log_FAR',
 'log_roadDensity',
 'log_intersectionDensity',
 'log_poiDensity',
 'log_maxHeight',
 'log_heightRange',
 'log_heightIndex',
 'log_heightDensity',
 'log_heightVariety',
 'log_cornerCountTotal',
 'log_basePerimeterMax',
 'log_basePerimeterMin',
 'log_areaVariance',
 'log_shape3DIndex',
 'log_balanceIndex',
 'log_cornerCountAvg',
 'log_buildingProximity',
 'log_buildingMinDist',
 'log_buildingDistVar',
 'log_streetRatio',
 'avgHeight',
 'compactness',
 'baseAreaTotal',
 'coverageRatio',
 'poiDiversity',
 'FVC',
 'permeableRatio',
 'minHeight',
 'shapeComplexity',
 'parcelArea',
 'largestPatchIndex',
 'buildingMaxDist',
 'buildingDistAvg',
 'SVF']

In [9]:
for col in skewed_cols:
    log_col = f'log_{col}'
    if log_col in grid500_transformed.columns:
        corr_before = grid500_transformed[col].corr(grid500_transformed['log_VIT202311'])
        corr_after = grid500_transformed[log_col].corr(grid500_transformed['log_VIT202311'])
        print(f"{col}: corr before = {corr_before:.3f}, after log = {corr_after:.3f}")

buildingCount: corr before = 0.479, after log = 0.675
totalHeight: corr before = 0.508, after log = 0.674
basePerimeterTotal: corr before = 0.625, after log = 0.641
basePerimeterAvg: corr before = 0.277, after log = 0.451
totalArea: corr before = 0.593, after log = 0.626
avgBuildingArea: corr before = 0.152, after log = 0.496
FAR: corr before = 0.593, after log = 0.656
roadDensity: corr before = 0.313, after log = 0.451
intersectionDensity: corr before = 0.230, after log = 0.454
poiDensity: corr before = 0.382, after log = 0.554
maxHeight: corr before = 0.521, after log = 0.604
heightRange: corr before = 0.499, after log = 0.613
heightIndex: corr before = -0.304, after log = -0.312
heightDensity: corr before = 0.508, after log = 0.510
heightVariety: corr before = 0.442, after log = 0.482
cornerCountTotal: corr before = 0.511, after log = 0.680
basePerimeterMax: corr before = 0.455, after log = 0.537
basePerimeterMin: corr before = -0.168, after log = 0.006
areaVariance: corr before = 0

# Model training

## Quick test

### performance indices

In [7]:
features = [
    'buildingCount', 'totalHeight', 'avgHeight', 'basePerimeterTotal',
    'basePerimeterAvg', 'compactness', 'baseAreaTotal', 'totalArea',
    'avgBuildingArea', 'FAR', 'coverageRatio',
    'roadDensity', 'intersectionDensity', 'poiDensity', 'poiDiversity',
    'FVC', 'permeableRatio', 'maxHeight', 'minHeight', 'heightRange', 'heightIndex',
    'heightDensity', 'heightVariety', 'cornerCountTotal', 'basePerimeterMax',
    'basePerimeterMin', 'shapeComplexity', 'areaVariance', 'parcelArea',
    'largestPatchIndex', 'shape3DIndex', 'balanceIndex', 'cornerCountAvg',
    'buildingProximity', 'buildingMinDist', 'buildingMaxDist',
    'buildingDistAvg', 'buildingDistVar', 'SVF', 'streetRatio'
]

targets = ['log_NTL2023_focal', 'log_VIT202311', 'UHIDAY2020_07_inv']

# 模型字典
models = {
    "XGBoost": lambda: XGBRegressor(objective='reg:squarederror', max_depth=5, learning_rate=0.1, n_estimators=100, seed=27),
}

for target in targets:
    print(f"\n=== Target: {target} ===")

    df_clean = grid500.dropna(subset=[target] + features).copy()
    print(df_clean.shape)

    X = df_clean[features]
    y = df_clean[target]

    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=45
    )

    for model_name, model_func in models.items():
        model = model_func()
        model.fit(X_train, y_train)
        y_pred = model.predict(X_test)

        mse = mean_squared_error(y_test, y_pred)
        r2 = r2_score(y_test, y_pred)

        print(f"\n{model_name} → R²: {r2:.4f}, MSE: {mse:.2f}")

        # 特征重要性（可选）
        feat_imp = pd.DataFrame({
            "feature": features,
            "importance": model.feature_importances_
        }).sort_values("importance", ascending=False)
        print(feat_imp.head(40))


=== Target: log_NTL2023_focal ===
(194941, 52)

XGBoost → R²: 0.4913, MSE: 0.21
                feature  importance
19          heightRange        0.30
28           parcelArea        0.14
14         poiDiversity        0.11
17            maxHeight        0.07
13           poiDensity        0.07
15                  FVC        0.06
12  intersectionDensity        0.06
32       cornerCountAvg        0.02
16       permeableRatio        0.02
20          heightIndex        0.01
11          roadDensity        0.01
18            minHeight        0.01
0         buildingCount        0.01
22        heightVariety        0.01
27         areaVariance        0.01
38                  SVF        0.01
39          streetRatio        0.01
3    basePerimeterTotal        0.01
8       avgBuildingArea        0.01
2             avgHeight        0.01
6         baseAreaTotal        0.01
4      basePerimeterAvg        0.01
29    largestPatchIndex        0.00
23     cornerCountTotal        0.00
25     basePerimete

### k=100, 300, 500 comparison

In [4]:
features = [
    'buildingCount', 'totalHeight', 'avgHeight', 'basePerimeterTotal',
    'basePerimeterAvg', 'compactness', 'baseAreaTotal', 'totalArea',
    'avgBuildingArea', 'FAR', 'coverageRatio',
    'roadDensity', 'intersectionDensity', 'poiDensity', 'poiDiversity',
    'FVC', 'permeableRatio', 'maxHeight', 'minHeight', 'heightRange', 'heightIndex',
    'heightDensity', 'heightVariety', 'cornerCountTotal', 'basePerimeterMax',
    'basePerimeterMin', 'shapeComplexity', 'areaVariance', 'parcelArea',
    'largestPatchIndex', 'shape3DIndex', 'balanceIndex', 'cornerCountAvg',
    'buildingProximity', 'buildingMinDist', 'buildingMaxDist',
    'buildingDistAvg', 'buildingDistVar', 'SVF', 'streetRatio'
]

# targets = ['log_NTL2023_focal', 'log_VIT202311', 'UHIDAY2020_07_inv', 'Efficiency_RBF']
# targets = ['Efficiency_RBF_k100', 'Efficiency_RBF_k300', 'Efficiency_RBF_k500']

# 模型字典
models = {
    "XGBoost": lambda: XGBRegressor(objective='reg:squarederror', max_depth=5, learning_rate=0.1, n_estimators=100, seed=27),
}

for target in targets:
    print(f"\n=== Target: {target} ===")

    df_clean = grid500.dropna(subset=[target] + features).copy()
    print(df_clean.shape)

    X = df_clean[features]
    y = df_clean[target]

    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42
    )

    for model_name, model_func in models.items():
        model = model_func()
        model.fit(X_train, y_train)
        y_pred = model.predict(X_test)

        mse = mean_squared_error(y_test, y_pred)
        r2 = r2_score(y_test, y_pred)

        print(f"\n{model_name} → R²: {r2:.4f}, MSE: {mse:.2f}")

        # 特征重要性（可选）
        feat_imp = pd.DataFrame({
            "feature": features,
            "importance": model.feature_importances_
        }).sort_values("importance", ascending=False)
        print(feat_imp.head(10))


=== Target: Efficiency_RBF_k100 ===
(194941, 53)

XGBoost → R²: 0.5283, MSE: 0.01
                feature  importance
3    basePerimeterTotal        0.38
7             totalArea        0.19
13  intersectionDensity        0.06
18            maxHeight        0.05
15         poiDiversity        0.04
12          roadDensity        0.03
39                  SVF        0.03
14           poiDensity        0.03
24     cornerCountTotal        0.03
20          heightRange        0.03

=== Target: Efficiency_RBF_k300 ===
(194941, 53)

XGBoost → R²: 0.4817, MSE: 0.01
                feature  importance
20          heightRange        0.26
15         poiDiversity        0.17
18            maxHeight        0.10
13  intersectionDensity        0.08
14           poiDensity        0.08
12          roadDensity        0.04
7             totalArea        0.03
16                  FVC        0.03
29           parcelArea        0.03
3    basePerimeterTotal        0.02

=== Target: Efficiency_RBF_k500 ===
(1949

### transformed features

In [12]:
grid500.head()

Global_ID     City  NTL2023  NTL2023_focal  VIT202311  UHIDAY2020_07_inv  \
0          0  Beijing     6.18           7.03    1649.86               0.00   
1          1  Beijing     5.74           7.23     155.86               0.00   
2          2  Beijing    11.32          13.62      57.14               0.00   
3          3  Beijing    11.73          14.36     243.57               0.00   
4          4  Beijing     8.45           9.19       0.00               0.00   

   buildingCount  totalHeight  avgHeight  maxHeight  minHeight  heightRange  \
0             31        93.00       3.00       3.00       3.00         0.00   
1             29        87.00       3.00       3.00       3.00         0.00   
2             22        66.00       3.00       3.00       3.00         0.00   
3              8        24.00       3.00       3.00       3.00         0.00   
4              0         0.00       0.00       0.00       0.00         0.00   

   heightIndex  heightDensity  heightVariety  cornerCountTotal  \
0         0.03           0.00           0.00               180   
1         0.03           0.00           0.00               149   
2         0.05           0.00           0.00               117   
3         0.12           0.00           0.00                42   
4         0.00           0.00           0.00                 0   

   basePerimeterTotal  basePerimeterAvg  basePerimeterMax  basePerimeterMin  \
0             6500.98            209.71            831.63             32.55   
1             3810.39            131.39            272.14             15.33   
2             3645.40            165.70            668.26             28.73   
3             1147.60            143.45            283.01             75.66   
4                0.00              0.00              0.00              0.00   

   shapeComplexity  compactness  baseAreaTotal  totalArea  areaVariance  \
0             5.07         0.56       88704.58   88704.58   50961857.70   
1             5.17         0.62       33119.49   33119.49    1261729.88   
2             4.92         0.56       51092.86   51092.86   30358745.83   
3             5.70         0.60        8534.61    8534.61     958399.89   
4             0.00         0.00           0.00       0.00          0.00   

   avgBuildingArea  parcelArea  largestPatchIndex  shape3DIndex  SVF  \
0          2861.44   132998.25               0.47          0.01 0.80   
1          1142.05   196625.22               0.67          0.26 0.89   
2          2322.40   132197.31               0.53          0.01 0.90   
3          1066.83   205685.37               0.82          0.02 0.95   
4             0.00       99.15               0.00          0.00 0.98   

   balanceIndex  FAR  coverageRatio  FVC  permeableRatio  poiDensity  \
0        234.60 0.35           0.35 0.84            0.89           4   
1         35.66 0.13           0.13 0.86            0.94          12   
2        151.50 0.20           0.20 0.88            0.93           0   
3         15.54 0.03           0.03 0.93            0.98           0   
4          0.00 0.00           0.00 0.82            0.98          16   

   poiDiversity  streetRatio  roadDensity  intersectionDensity  \
0          0.00         0.20            5                    1   
1          0.64         0.20            7                    1   
2          0.00         0.20            1                    0   
3          0.00         0.20            3                    1   
4          0.69         0.00            1                    1   

   cornerCountAvg  buildingProximity  buildingMinDist  buildingMaxDist  \
0            5.81              52.56            14.99           620.94   
1            5.14              34.22             6.82           594.16   
2            5.32              47.90            17.34           610.84   
3            5.25              35.93            13.49           474.17   
4            0.00               0.00             0.00             0.00   

   buildingDistAvg  bu

In [14]:
targets = ['log_NTL2023_focal', 'log_VIT202311', 'UHIDAY2020_07_inv', 'Efficiency_RBF_k100', 'Efficiency_RBF_k500']
# targets = ['Efficiency_RBF_k100', 'Efficiency_RBF_k300', 'Efficiency_RBF_k500']

# 模型字典
models = {
    "XGBoost": lambda: XGBRegressor(objective='reg:squarederror', max_depth=5, learning_rate=0.1, n_estimators=100, seed=27),
}

for target in targets:
    print(f"\n=== Target: {target} ===")

    df_clean = grid500_transformed.dropna(subset=[target] + features_transformed).copy()
    print(df_clean.shape)

    X = df_clean[features_transformed]
    y = df_clean[target]

    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42
    )

    for model_name, model_func in models.items():
        model = model_func()
        model.fit(X_train, y_train)
        y_pred = model.predict(X_test)

        mse = mean_squared_error(y_test, y_pred)
        r2 = r2_score(y_test, y_pred)

        print(f"\n{model_name} → R²: {r2:.4f}, MSE: {mse:.2f}")

        # 特征重要性（可选）
        feat_imp = pd.DataFrame({
            "feature": features_transformed,
            "importance": model.feature_importances_
        }).sort_values("importance", ascending=False)
        print(feat_imp.head(10))


=== Target: log_NTL2023_focal ===
(194941, 77)

XGBoost → R²: 0.4839, MSE: 0.21
                    feature  importance
11          log_heightRange        0.25
35               parcelArea        0.16
30             poiDiversity        0.11
9            log_poiDensity        0.09
31                      FVC        0.07
10            log_maxHeight        0.07
8   log_intersectionDensity        0.06
21       log_cornerCountAvg        0.02
32           permeableRatio        0.02
12          log_heightIndex        0.01

=== Target: log_VIT202311 ===
(194941, 77)

XGBoost → R²: 0.6860, MSE: 1.14
                    feature  importance
2    log_basePerimeterTotal        0.43
4             log_totalArea        0.12
28            baseAreaTotal        0.09
38          buildingDistAvg        0.04
11          log_heightRange        0.04
39                      SVF        0.04
37          buildingMaxDist        0.03
10            log_maxHeight        0.03
8   log_intersectionDensity        0.03
30

KeyError: ['Efficiency_RBF_k100']

In [15]:
combined_features = [f'log_{col}' for col in skewed_cols] + skewed_cols + non_log_features
len(combined_features)

66

In [16]:
# targets = ['log_NTL2023_focal', 'log_VIT202311', 'UHIDAY2020_07_inv', 'Efficiency_RBF_k100', 'Efficiency_RBF_k500']
# targets = ['Efficiency_RBF_k100', 'Efficiency_RBF_k300', 'Efficiency_RBF_k500']
targets = ['log_VIT202311']

# 模型字典
models = {
    "XGBoost": lambda: XGBRegressor(objective='reg:squarederror', max_depth=5, learning_rate=0.1, n_estimators=100, seed=27),
}

for target in targets:
    print(f"\n=== Target: {target} ===")

    df_clean = grid500_transformed.copy()
    print(df_clean.shape)

    X = df_clean[combined_features]
    y = df_clean[target]

    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42
    )

    for model_name, model_func in models.items():
        model = model_func()
        model.fit(X_train, y_train)
        print(X_train.shape)
        y_pred = model.predict(X_test)

        mse = mean_squared_error(y_test, y_pred)
        r2 = r2_score(y_test, y_pred)

        print(f"\n{model_name} → R²: {r2:.4f}, MSE: {mse:.2f}")

        # 特征重要性（可选）
        feat_imp = pd.DataFrame({
            "feature": combined_features,
            "importance": model.feature_importances_
        }).sort_values("importance", ascending=False)
        print(feat_imp.head(68))


=== Target: log_VIT202311 ===
(194941, 77)
(155952, 66)

XGBoost → R²: 0.6860, MSE: 1.14
                   feature  importance
2   log_basePerimeterTotal        0.43
4            log_totalArea        0.11
54           baseAreaTotal        0.09
64         buildingDistAvg        0.04
11         log_heightRange        0.04
..                     ...         ...
40           heightVariety        0.00
39           heightDensity        0.00
38             heightIndex        0.00
37             heightRange        0.00
33             roadDensity        0.00

[66 rows x 2 columns]


### Conclusion: Log-transforming predictors did not improve model R² and MSE
- The XGBoost model (log_VIT202311) achieved the same R² (0.6860) and MSE (1.14) across all tested versions:
    - Using original features
    - Using log-transformed features
    - Using both original + log-transformed features
- Although log transformation improved the correlation between individual features and the target variable, it did not enhance model performance.
- This is because XGBoost is a non-linear, tree-based model that can naturally handle skewed or non-normal feature distributions without needing transformations.
- Therefore, for this task and model type, log-transforming the predictors is not necessary from a predictive performance standpoint.

## Model comparison

In [13]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from xgboost import XGBRegressor
import pandas as pd

# Feature and target definitions
features = [
    'buildingCount', 'totalHeight', 'avgHeight', 'basePerimeterTotal',
    'basePerimeterAvg', 'compactness', 'baseAreaTotal', 'totalArea',
    'avgBuildingArea', 'FAR', 'coverageRatio', 'roadDensity',
    'intersectionDensity', 'poiDensity', 'poiDiversity', 'FVC',
    'permeableRatio', 'maxHeight', 'minHeight', 'heightRange',
    'heightIndex', 'heightDensity', 'heightVariety', 'cornerCountTotal',
    'basePerimeterMax', 'basePerimeterMin', 'shapeComplexity',
    'areaVariance', 'parcelArea', 'largestPatchIndex', 'shape3DIndex',
    'balanceIndex', 'cornerCountAvg', 'buildingProximity',
    'buildingMinDist', 'buildingMaxDist', 'buildingDistAvg',
    'buildingDistVar', 'SVF', 'streetRatio'
]

targets = ['log_NTL2023_focal', 'log_VIT202311', 'UHIDAY2020_07_inv']

models = {
    "XGBoost": lambda: XGBRegressor(objective='reg:squarederror', max_depth=5, learning_rate=0.1, n_estimators=100, seed=27),
    "RandomForest": lambda: RandomForestRegressor(n_estimators=50, max_depth=6, random_state=27),
    "GradientBoosting": lambda: GradientBoostingRegressor(n_estimators=50, learning_rate=0.1, max_depth=5, random_state=27),
}

# Store all results
all_results = []

for target in targets:
    df_clean = grid500.dropna(subset=[target] + features).copy()
    X = df_clean[features]
    y = df_clean[target]

    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=45
    )

    for model_name, model_func in models.items():
        model = model_func()
        model.fit(X_train, y_train)

        y_train_pred = model.predict(X_train)
        y_test_pred = model.predict(X_test)

        r2_train = r2_score(y_train, y_train_pred)
        r2_test = r2_score(y_test, y_test_pred)
        print(r2_test)
        
        rmse_train = mean_squared_error(y_train, y_train_pred, squared=False)
        rmse_test = mean_squared_error(y_test, y_test_pred, squared=False)

        all_results.append({
            "target": target,
            "model": model_name,
            "r2_train": r2_train,
            "r2_test": r2_test,
            "rmse_train": rmse_train,
            "rmse_test": rmse_test
        })

# Convert to DataFrame for summary
results_df = pd.DataFrame(all_results)
results_df = results_df.sort_values(by=["target", "r2_test"], ascending=[True, False])

results_df

0.4913203442619388
0.417643952575137
0.47685280637379834
0.6878740558898829
0.6420935600905928
0.6818949409499258
0.40843577757431526
0.3376794931857777
0.3899382458565389


target             model  r2_train  r2_test  rmse_train  \
6  UHIDAY2020_07_inv           XGBoost      0.43     0.41        0.76   
8  UHIDAY2020_07_inv  GradientBoosting      0.40     0.39        0.77   
7  UHIDAY2020_07_inv      RandomForest      0.34     0.34        0.81   
0  log_NTL2023_focal           XGBoost      0.51     0.49        0.45   
2  log_NTL2023_focal  GradientBoosting      0.49     0.48        0.46   
1  log_NTL2023_focal      RandomForest      0.43     0.42        0.49   
3      log_VIT202311           XGBoost      0.70     0.69        1.04   
5      log_VIT202311  GradientBoosting      0.69     0.68        1.06   
4      log_VIT202311      RandomForest      0.64     0.64        1.13   

   rmse_test  
6       0.77  
8       0.78  
7       0.81  
0       0.46  
2       0.47  
1       0.49  
3       1.06  
5       1.07  
4       1.13

In [14]:
results_df.to_csv(r"../output/model_comparison.csv")

### Using AutoGluon (TBC)

In [11]:
from autogluon.tabular import TabularPredictor
from sklearn.model_selection import train_test_split
import pandas as pd

features = [
    'buildingCount', 'totalHeight', 'avgHeight', 'basePerimeterTotal',
    'basePerimeterAvg', 'compactness', 'baseAreaTotal', 'totalArea',
    'avgBuildingArea', 'FAR', 'coverageRatio',
    'roadDensity', 'intersectionDensity', 'poiDensity', 'poiDiversity',
    'FVC', 'permeableRatio', 'maxHeight', 'minHeight', 'heightRange', 'heightIndex',
    'heightDensity', 'heightVariety', 'cornerCountTotal', 'basePerimeterMax',
    'basePerimeterMin', 'shapeComplexity', 'areaVariance', 'parcelArea',
    'largestPatchIndex', 'shape3DIndex', 'balanceIndex', 'cornerCountAvg',
    'buildingProximity', 'buildingMinDist', 'buildingMaxDist',
    'buildingDistAvg', 'buildingDistVar', 'SVF', 'streetRatio'
]

targets = ['log_NTL2023_focal', 'log_VIT202311', 'UHIDAY2020_07_inv']

for target in targets:
    print(f"\nTraining for target: {target}")
    
    df_clean = grid500.dropna(subset=[target] + features).copy()
    df_clean = df_clean[features + [target]]
    
    # Split train-test for evaluation
    train_data, test_data = train_test_split(df_clean, test_size=0.2, random_state=42)
    
    # Define predictor
    predictor = TabularPredictor(label=target, eval_metric='r2', verbosity=2).fit(train_data)

    # Evaluate on test data
    performance = predictor.evaluate(test_data)
    
    # 输出排名靠前的模型
    leaderboard = predictor.leaderboard(test_data, silent=True)
    leaderboard

ModuleNotFoundError: No module named 'autogluon'

In [ ]:
leaderboard.to_csv("../output/model_comparison_autogluon.csv)

## XGBoost training

### Predictors: 41 urban form indicators

In [17]:
grid500.describe()

Global_ID   NTL2023  NTL2023_focal  VIT202311  UHIDAY2020_07_inv  \
count  194941.00 189742.00      194941.00  194941.00          194941.00   
mean    98137.51     28.88          28.65    1414.14              -0.99   
std     56862.94     19.36          17.87    1821.23               1.00   
min         0.00      0.00           0.00       0.00              -4.60   
25%     49033.00     15.06          15.64     206.86              -1.69   
50%     97800.00     25.85          26.01     731.14              -0.91   
75%    147135.00     38.67          38.17    2003.43               0.00   
max    196769.00    465.96         348.37   68494.71               5.13   

       buildingCount  totalHeight  avgHeight  maxHeight  minHeight  \
count      194941.00    194941.00  194941.00  194941.00  194941.00   
mean           46.70       893.91      16.53      30.42       9.58   
std            50.96      1035.48       8.72      24.26       4.91   
min             0.00         0.00       0.00       0.00       0.00   
25%            12.00       144.00      12.00      12.00       6.00   
50%            35.00       559.00      16.67      27.00      12.00   
75%            65.00      1315.00      21.99      40.00      12.00   
max          1189.00     22354.50     130.50     535.00     112.00   

       heightRange  heightIndex  heightDensity  heightVariety  \
count    194941.00    194941.00      194941.00      194941.00   
mean         20.83         0.10           0.00           0.30   
std          23.99         0.17           0.00           0.28   
min           0.00         0.00           0.00           0.00   
25%           0.00         0.02           0.00           0.00   
50%          15.00         0.04           0.00           0.28   
75%          28.00         0.09           0.01           0.43   
max         532.00         1.00           0.09           2.87   

       cornerCountTotal  basePerimeterTotal  basePerimeterAvg  \
count         194941.00           194941.00         194941.00   
mean             315.42             4959.32            112.06   
std              352.35             4245.40             60.43   
min                0.00                0.00              0.00   
25%               72.00             1417.38             83.02   
50%              222.00             4200.02            110.75   
75%              445.00             7512.92            137.30   
max             8619.00            50912.47           1689.25   

       basePerimeterMax  basePerimeterMin  shapeComplexity  compactness  \
count         194941.00         194941.00        194941.00    194941.00   
mean             347.82             22.65             4.55         0.54   
std              230.34             35.73             1.40         0.17   
min                0.00              0.00             0.00         0.00   
25%              202.45              4.35             4.57         0.53   
50%              319.04             14.07             4.82         0.57   
75%              461.66             28.38             5.08         0.62   
max             5146.21           1689.25            39.49         0.99   

       baseAreaTotal  totalArea    areaVariance  avgBuildingArea  parcelArea  \
count      194941.00  194941.00       194941.00        194941.00   194941.00   
mean        33192.68  216494.27    191423545.70          5437.01   180006.93   
std         27474.13  210010.02   2300870309.71          7776.02    80149.49   
min             0.00       0.00            0.00             0.00        0.00   
25%          8888.17   37241.51      3519241.57          1875.57   131267.14   
50%         29057.71  157756.72     20726892.83          4027.77   222107.39   
75%         51784.69  347326.83     74777609.98          6726.97   238506.86   
max        226640.62 3672930.95 662000000000.00        664544.83   250000.00   

       largestPatchIndex  shape3DIndex       SVF  balanceIndex       FAR  \
count          194941.00     194941.00 194941.00     19

In [18]:
# Set global random seed for reproducibility
SEED = 42
np.random.seed(SEED)

# Ensure output directories exist
os.makedirs("../output/models", exist_ok=True)
os.makedirs("../output/param_logs", exist_ok=True)

# ========================= XGBoost Training Loop =========================
targets = ['log_NTL2023_focal', 'log_VIT202311', 'UHIDAY2020_07_inv']

for target in targets:
    print(f"\n===== Training model for target: {target} =====")

    # Filter rows with non-missing values for target and features
    df_target = grid500.copy()
    
    # Split features and target
    X = df_target[features]
    y = df_target[target]

    # Train-test split
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=SEED
    )

    print(f"Training samples: {X_train.shape[0]}, Testing samples: {X_test.shape[0]}")

    # Define base XGBoost model
    xgb_model = xgb.XGBRegressor(
        objective='reg:squarederror',
        random_state=SEED,
        n_jobs=-1
    )

    # Randomized hyperparameter search
    param_dist = {
    'n_estimators': [100, 200, 300, 500],
    'max_depth': [3, 5, 7, 9],
    'learning_rate': [0.01, 0.03, 0.05, 0.07, 0.1],
    'subsample': [0.7, 0.9, 1.0],
    'colsample_bytree': [0.7, 0.9, 1.0],
    'gamma': [0, 0.1, 0.3],
    'reg_alpha': [0, 0.1, 0.5],
    'reg_lambda': [1, 1.5, 2],
    'min_child_weight': [1, 3, 5],
    }
    
    random_search = RandomizedSearchCV(
        estimator=xgb_model,
        param_distributions=param_dist,
        n_iter=200,
        cv=5,
        scoring='neg_mean_squared_error',
        verbose=1,
        random_state=SEED
    )

    # Fit model
    random_search.fit(X_train, y_train)

    # Retrieve best estimator
    best_model = random_search.best_estimator_

    # Save trained model
    model_path = f"../output/models/{target}_xgb_model.pkl"
    joblib.dump(best_model, model_path)
    print(f"✔ Saved best model to: {model_path}")

    # Save best hyperparameters
    best_params = random_search.best_params_
    param_path = f"../output/param_logs/{target}_best_params.json"
    with open(param_path, 'w') as f:
        json.dump(best_params, f, indent=4)
    print(f"✔ Saved best parameters to: {param_path}")

    # Evaluate model performance
    y_pred_train = best_model.predict(X_train)
    y_pred_test = best_model.predict(X_test)

    train_rmse = mean_squared_error(y_train, y_pred_train, squared=False)
    test_rmse = mean_squared_error(y_test, y_pred_test, squared=False)
    test_r2 = r2_score(y_test, y_pred_test)

    print(f"Train RMSE: {train_rmse:.4f}")
    print(f"Test  RMSE: {test_rmse:.4f}")
    print(f"Test   R² : {test_r2:.4f}")

print("\n✅ All models trained successfully.")


===== Training model for target: log_NTL2023_focal =====
Training samples: 155952, Testing samples: 38989
Fitting 5 folds for each of 200 candidates, totalling 1000 fits
✔ Saved best model to: ../output/models/log_NTL2023_focal_xgb_model.pkl
✔ Saved best parameters to: ../output/param_logs/log_NTL2023_focal_best_params.json
Train RMSE: 0.3570
Test  RMSE: 0.4478
Test   R² : 0.5176

===== Training model for target: log_VIT202311 =====
Training samples: 155952, Testing samples: 38989
Fitting 5 folds for each of 200 candidates, totalling 1000 fits
✔ Saved best model to: ../output/models/log_VIT202311_xgb_model.pkl
✔ Saved best parameters to: ../output/param_logs/log_VIT202311_best_params.json
Train RMSE: 0.8110
Test  RMSE: 1.0461
Test   R² : 0.6982

===== Training model for target: UHIDAY2020_07_inv =====
Training samples: 155952, Testing samples: 38989
Fitting 5 folds for each of 200 candidates, totalling 1000 fits
✔ Saved best model to: ../output/models/UHIDAY2020_07_inv_xgb_model.pkl
✔

### Predictors: 41 urban form indicators + one-hot city columns (TBC)

In [6]:
# One-hot encode the 'City' column
city_encoder = OneHotEncoder(sparse=False, handle_unknown='ignore')
city_onehot = city_encoder.fit_transform(grid500[['City']])
city_feature_names = city_encoder.get_feature_names_out(['City'])

# Create one-hot DataFrame and concatenate
df_city_encoded = pd.DataFrame(city_onehot, columns=city_feature_names, index=grid500.index)
grid500_with_city = pd.concat([grid500, df_city_encoded], axis=1)
grid500_with_city.describe()

Global_ID   NTL2023  NTL2023_focal  VIT202311  UHIDAY2020_07_inv  \
count  194941.00 189742.00      194941.00  194941.00          194941.00   
mean    98137.51     28.88          28.65    1414.14              -0.99   
std     56862.94     19.36          17.87    1821.23               1.00   
min         0.00      0.00           0.00       0.00              -4.60   
25%     49033.00     15.06          15.64     206.86              -1.69   
50%     97800.00     25.85          26.01     731.14              -0.91   
75%    147135.00     38.67          38.17    2003.43               0.00   
max    196769.00    465.96         348.37   68494.71               5.13   

       buildingCount  totalHeight  avgHeight  maxHeight  minHeight  \
count      194941.00    194941.00  194941.00  194941.00  194941.00   
mean           46.70       893.91      16.53      30.42       9.58   
std            50.96      1035.48       8.72      24.26       4.91   
min             0.00         0.00       0.00       0.00       0.00   
25%            12.00       144.00      12.00      12.00       6.00   
50%            35.00       559.00      16.67      27.00      12.00   
75%            65.00      1315.00      21.99      40.00      12.00   
max          1189.00     22354.50     130.50     535.00     112.00   

       heightRange  heightIndex  heightDensity  heightVariety  \
count    194941.00    194941.00      194941.00      194941.00   
mean         20.83         0.10           0.00           0.30   
std          23.99         0.17           0.00           0.28   
min           0.00         0.00           0.00           0.00   
25%           0.00         0.02           0.00           0.00   
50%          15.00         0.04           0.00           0.28   
75%          28.00         0.09           0.01           0.43   
max         532.00         1.00           0.09           2.87   

       cornerCountTotal  basePerimeterTotal  basePerimeterAvg  \
count         194941.00           194941.00         194941.00   
mean             315.42             4959.32            112.06   
std              352.35             4245.40             60.43   
min                0.00                0.00              0.00   
25%               72.00             1417.38             83.02   
50%              222.00             4200.02            110.75   
75%              445.00             7512.92            137.30   
max             8619.00            50912.47           1689.25   

       basePerimeterMax  basePerimeterMin  shapeComplexity  compactness  \
count         194941.00         194941.00        194941.00    194941.00   
mean             347.82             22.65             4.55         0.54   
std              230.34             35.73             1.40         0.17   
min                0.00              0.00             0.00         0.00   
25%              202.45              4.35             4.57         0.53   
50%              319.04             14.07             4.82         0.57   
75%              461.66             28.38             5.08         0.62   
max             5146.21           1689.25            39.49         0.99   

       baseAreaTotal  totalArea    areaVariance  buildingIntensity  \
count      194941.00  194941.00       194941.00          194941.00   
mean        33192.68  216494.27    191423545.70               0.87   
std         27474.13  210010.02   2300870309.71               0.84   
min             0.00       0.00            0.00               0.00   
25%          8888.17   37241.51      3519241.57               0.15   
50%         29057.71  157756.72     20726892.83               0.63   
75%         51784.69  347326.83     74777609.98               1.39   
max        226640.62 3672930.95 662000000000.00              14.69   

       avgBuildingArea  parcelArea  largestPatchIndex  shape3DIndex       SVF  \
count        194941.00   194941.00          194941.00     194941.00 194941.00   
mean           5437.01   180006.93               0.50      

In [7]:
grid500_with_city.to_csv("../data/grid500_36cities_with_eff_city.csv")

In [48]:
# Set global random seed for reproducibility
SEED = 42
np.random.seed(SEED)

# Ensure output directories exist
os.makedirs("../output/models_with_city", exist_ok=True)
os.makedirs("../output/param_logs_with_city", exist_ok=True)

# One-hot encode the 'City' column
city_encoder = OneHotEncoder(sparse=False, handle_unknown='ignore')
city_onehot = city_encoder.fit_transform(grid500[['City']])
city_feature_names = city_encoder.get_feature_names_out(['City'])

# Create one-hot DataFrame and concatenate
df_city_encoded = pd.DataFrame(city_onehot, columns=city_feature_names, index=grid500.index)
grid500_with_city = pd.concat([grid500, df_city_encoded], axis=1)

# ========================= XGBoost Training Loop =========================
targets = ['log_NTL2023_focal', 'log_VIT202311', 'UHIDAY2020_07_inv']

# Combine city features with existing morphological features
features_extended = features + list(city_feature_names)

for target in targets:
    print(f"\n===== Training model for target: {target} =====")

    # Drop rows with missing target or features
    df_target = grid500_with_city.dropna(subset=features_extended + [target])

    # Split features and target
    X = df_target[features_extended]
    y = df_target[target]

    # Train-test split
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=SEED
    )

    print(f"Training samples: {X_train.shape[0]}, Testing samples: {X_test.shape[0]}")

    # Define base XGBoost model
    xgb_model = xgb.XGBRegressor(
        objective='reg:squarederror',
        random_state=SEED,
        n_jobs=-1
    )

    # Randomized hyperparameter search
    param_dist = {
        'n_estimators': [100, 200, 300, 500],
        'max_depth': [3, 5, 7, 9],
        'learning_rate': [0.01, 0.03, 0.05, 0.07, 0.1],
        'subsample': [0.7, 0.9, 1.0],
        'colsample_bytree': [0.7, 0.9, 1.0],
        'gamma': [0, 0.1, 0.3],
        'reg_alpha': [0, 0.1, 0.5],
        'reg_lambda': [1, 1.5, 2],
        'min_child_weight': [1, 3, 5],
    }

    random_search = RandomizedSearchCV(
        estimator=xgb_model,
        param_distributions=param_dist,
        n_iter=200,
        cv=5,
        scoring='neg_mean_squared_error',
        verbose=1,
        random_state=SEED
    )

    # Fit model
    random_search.fit(X_train, y_train)

    # Retrieve best estimator
    best_model = random_search.best_estimator_

    # Save trained model
    model_path = f"../output/models_with_city/{target}_xgb_model.pkl"
    joblib.dump(best_model, model_path)
    print(f"✔ Saved best model to: {model_path}")

    # Save best hyperparameters
    best_params = random_search.best_params_
    param_path = f"../output/param_logs_with_city/{target}_best_params.json"
    with open(param_path, 'w') as f:
        json.dump(best_params, f, indent=4)
    print(f"✔ Saved best parameters to: {param_path}")

    # Evaluate model performance
    y_pred_train = best_model.predict(X_train)
    y_pred_test = best_model.predict(X_test)

    train_rmse = mean_squared_error(y_train, y_pred_train, squared=False)
    test_rmse = mean_squared_error(y_test, y_pred_test, squared=False)
    test_r2 = r2_score(y_test, y_pred_test)

    print(f"Train RMSE: {train_rmse:.4f}")
    print(f"Test  RMSE: {test_rmse:.4f}")
    print(f"Test   R² : {test_r2:.4f}")

print("\n✅ All models trained successfully with city one-hot encoding.")


===== Training model for target: log_NTL2023_focal =====
Training samples: 155952, Testing samples: 38989
Fitting 5 folds for each of 200 candidates, totalling 1000 fits
✔ Saved best model to: ../output/models_with_city/log_NTL2023_focal_xgb_model.pkl
✔ Saved best parameters to: ../output/param_logs_with_city/log_NTL2023_focal_best_params.json
Train RMSE: 0.2694
Test  RMSE: 0.3899
Test   R² : 0.6343

===== Training model for target: log_VIT202311 =====
Training samples: 155952, Testing samples: 38989
Fitting 5 folds for each of 200 candidates, totalling 1000 fits
✔ Saved best model to: ../output/models_with_city/log_VIT202311_xgb_model.pkl
✔ Saved best parameters to: ../output/param_logs_with_city/log_VIT202311_best_params.json
Train RMSE: 0.7999
Test  RMSE: 1.0103
Test   R² : 0.7186

===== Training model for target: UHIDAY2020_07_inv =====
Training samples: 155952, Testing samples: 38989
Fitting 5 folds for each of 200 candidates, totalling 1000 fits
✔ Saved best model to: ../output/

### one-hot city encoding
- Performance improved for all three targets after including city-level one-hot features.
- Urban-level heterogeneity matters, especially for environmental and vitality indicators.
- City dummy variables act as latent factors for unobservable systemic differences.

In [8]:
model_eva = pd.read_csv('../output/model_evaluation_summary.csv')
model_eva

41 urban form indicators + city one-hot encoding  Unnamed: 1 Unnamed: 2  \
0                                           Target  Train_RMSE  Test_RMSE   
1                                log_NTL2023_focal      0.2694     0.3899   
2                                    log_VIT202311      0.7999     1.0103   
3                                UHIDAY2020_07_inv      0.3949     0.5609   
4                                              NaN         NaN        NaN   
5                         41 urban form indicators         NaN        NaN   
6                                           Target  Train_RMSE  Test_RMSE   
7                                log_NTL2023_focal       0.357     0.4478   
8                                    log_VIT202311      0.9067     1.0483   
9                                UHIDAY2020_07_inv      0.5878     0.7378   

  Unnamed: 3  
0    Test_R2  
1     0.6343  
2     0.7186  
3     0.6844  
4        NaN  
5        NaN  
6    Test_R2  
7     0.5176  
8      0.697  
9      0.454